modules/ipp/scanner.go

```diff
@@ -0,0 +1,206 @@
+	//TODO: Write a help string
+	return ""
+}
+
+// Init initializes the Scanner.
+func (scanner *Scanner) Init(flags zgrab2.ScanFlags) error {
+	f, _ := flags.(*Flags)
+	scanner.config = f
+	//TODO: Take action in response to flags which were set
+	return nil
+}
+
+// InitPerSender initializes the scanner for a given sender.
+func (scanner *Scanner) InitPerSender(senderID int) error {
+	return nil
+}
+
+// GetName returns the Scanner name defined in the Flags.
+func (scanner *Scanner) GetName() string {
+	return scanner.config.Name
+}
+
+// Protocol returns the protocol identifier of the scan.
+func (scanner *Scanner) Protocol() string {
+	return "ipp"
+}
+
+// GetPort returns the port being scanned.
+func (scanner *Scanner) GetPort() uint {
+	return scanner.config.Port
+}
+
+//FIXME: Maybe switch to ipp/ipps schemes, at least optionally
+func getIPPURL(https bool, host string, port uint16, endpoint string) string {
+	var proto string
+	if https {
+		proto = "https"
+	} else {
+		proto = "http"
+	}
+	return proto + "://" + host + ":" + strconv.FormatUint(uint64(port), 10) + endpoint
+}
+
+func ippInContentType(resp http.Response) bool {
+	for _, t := range resp.Header["Content-Type"] {
+		if strings.Contains(t, CONTENT_TYPE) {
+			return true
+		}
+	}
+	return false
+}
+
+//TODO: Doesn't support TLS at all right now
+func (scanner *Scanner) grab(target zgrab2.ScanTarget) (int8, int8, *zgrab2.ScanError) {
+	//FIXME: This is not where this hostname assignment logic should live
+	host := target.Domain
+	if host == "" {
+		host = target.IP.String()
+	}
+	//FIXME: ?Should just use endpoint "/", since we get the same response as "/ipp" on CUPS??
+	uri := getIPPURL(scanner.config.IPPSecure, host, uint16(scanner.config.BaseFlags.Port), "/ipp")
+	b := getPrinterAttributesRequest(uri)
+	resp, err := http.Post(uri, CONTENT_TYPE, &b)
+	if err != nil {
+		//TODO: Create a descriptive error
+		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	if resp != nil && resp.Body != nil {
+		defer resp.Body.Close()
+	} else {
+		//TODO: Determine whether we need this error to avoid reading from Body
+		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	//FIXME: Maybe add something to handle redirects
+	//FIXME: Probably return the whole response for further inspection by ztag, rather
+	//         than grabbing first 2 bytes. In that case, implement maxRead like http module
+
+	//Check to make sure that the repsonse received is actually IPP
+	//Content-Type header matches is sufficient
+	//HTTP on port 631 is sufficient
+	//Still record data in the case of protocol error to see what that data looks like
+
+	//TODO: Record server-header version numbers
+	//protocols := resp.Header["Server"])
+	var version int16
+	if err := binary.Read(resp.Body, binary.BigEndian, &version); err != nil {
+		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	return int8(version >> 8), int8(version & 0xff), nil
```

Comment:
Should these be `var version uint16` / `uint8(version >> 8)` etc? Or are negative versions allowed?

Reply:
I don't believe that negative values are ever used, but the RFC specifies that _"Every integer in this encoding MUST be encoded as a signed integer using two's-complement binary encoding with big-endian format"_ (Source: https://tools.ietf.org/html/rfc8010#section-3)

Reply:
Got it -- I'd say just drop that link / sentence into a comment.

Also, I don't recall off-hand how `>>` works with signed integers (whether it keeps / extends the sign), but to be safe / explicit, it should probably be two `Read`s (of `int8`s), or change `version` to a `uint16` so that there is no ambiguity (since version is just being treated as a two-byte buffer anyway).